service/src/main/java/bio/terra/externalcreds/controllers/OauthApiController.java

```diff
@@ -0,0 +1,37 @@
+package bio.terra.externalcreds.controllers;
+
+import bio.terra.externalcreds.generated.api.OauthApi;
+import bio.terra.externalcreds.generated.model.Provider;
+import bio.terra.externalcreds.services.*;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.servlet.http.HttpServletRequest;
+import lombok.SneakyThrows;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+
+@Controller
+public record OauthApiController(
+    HttpServletRequest request,
+    ObjectMapper mapper,
+    ProviderService providerService,
+    ExternalCredsSamUserFactory samUserFactory)
+    implements OauthApi {
+
+  @Override
+  public ResponseEntity<String> getAuthUrl(Provider providerName, String redirectUri) {
+    var samUser = samUserFactory.from(request);
+
+    var authorizationUrl =
+        providerService.getProviderAuthorizationUrl(
+            samUser.getSubjectId(), providerName.toString(), redirectUri);
+
+    return ResponseEntity.of(authorizationUrl.map(this::jsonString));
```

Comment:
I'd like to take this opportunity to move away from returning JSON Strings. Since there's no structure to this response, encoding it as a JSON String buys us nothing. `text/plain` would be fine.